aptos-move/aptos-vm-types/src/module_and_script_storage/temporary_module_storage.rs

```diff
@@ -0,0 +1,138 @@
+    account_address::AccountAddress, identifier::IdentStr, metadata::Metadata,
+    vm_status::StatusCode,
+};
+use move_vm_runtime::{Module, ModuleStorage};
+use std::{collections::BTreeMap, sync::Arc};
+
+/// A module storage with stashed temporary changes (write ops). Use by AptosVM to
+/// process init_module, where published modules are made temporarily visible.
+pub struct TemporaryModuleStorage<'a, M> {
+    deserializer_config: &'a DeserializerConfig,
+
+    // TODO(loader_v2): Implement cache which "caches" deserialization? Or we can construct
+    //                  Arc<Module> and store it inside MVHashMap instead!
+    write_ops: BTreeMap<StateKey, WriteOp>,
+    module_storage: &'a M,
+}
+
+impl<'a, M: ModuleStorage> TemporaryModuleStorage<'a, M> {
+    /// Creates a new temporary module storage with stashed changes.
+    pub fn create(
+        deserializer_config: &'a DeserializerConfig,
+        write_ops: BTreeMap<StateKey, WriteOp>,
+        module_storage: &'a M,
+    ) -> Self {
+        Self {
+            deserializer_config,
+            write_ops,
+            module_storage,
+        }
+    }
+
+    /// Destroys temporary module storage releasing the stshed changes.
+    pub fn destroy(self) -> ModuleWriteSet {
+        // We do not care here about writes to special addresses because there is no flushing.
+        ModuleWriteSet::new(false, self.write_ops)
+    }
+
+    fn write_op_to_compiled_module(
+        &self,
+        write_op: &WriteOp,
+    ) -> PartialVMResult<Arc<CompiledModule>> {
+        // Modules can never be deleted, return an invariant violation for extra safety.
+        let bytes = write_op.bytes().ok_or_else(|| {
+            PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                .with_message("Module deletion is not be possible".to_string())
+        })?;
+        Ok(Arc::new(CompiledModule::deserialize_with_config(
+            bytes,
+            self.deserializer_config,
+        )?))
+    }
+}
+
+impl<'a, M: ModuleStorage> ModuleStorage for TemporaryModuleStorage<'a, M> {
+    fn check_module_exists(
+        &self,
+        address: &AccountAddress,
+        module_name: &IdentStr,
+    ) -> PartialVMResult<bool> {
+        let state_key = StateKey::module(address, module_name);
+        if self.write_ops.contains_key(&state_key) {
+            return Ok(true);
+        }
+        self.module_storage
+            .check_module_exists(address, module_name)
+    }
+
+    fn fetch_module_size_in_bytes(
+        &self,
+        address: &AccountAddress,
+        module_name: &IdentStr,
+    ) -> PartialVMResult<usize> {
+        let state_key = StateKey::module(address, module_name);
+        match self.write_ops.get(&state_key) {
+            Some(write_op) => Ok(write_op.size()),
+            None => self
+                .module_storage
+                .fetch_module_size_in_bytes(address, module_name),
+        }
+    }
+
+    fn fetch_module_metadata(
```

Comment:
Are we concerned about the performance impact here? Seems that we would be deserializing the write op bytes at all time.

Reply:
We should be, but we can add caching later not in this PR, here I just add an interface, I think I had a TODO somewhere for `Module` construction.

Reply:
A few notes about these:
- See #14395: the temporary storage is move to Move, so no need to process write ops, and there I use `UnsyncModuleStorage` to cache verification/deserialization. - Right now all metadata calls are cloning and so are the ones in `MoveResolver` (`get_module_metadata`) - I think there are a few cases where you construct a module, yet unowned, and have to give out a reference. We should definitely use `&[..]` but this can be done once everything is connected?

Reply:
One other interface I wanted is to query neighbours (dependencies and friends), but right now it gets ugly because of arcs..